server/webapp/WEB-INF/rails/webpack/views/pages/pipeline_configs/materials/index.tsx

```diff
@@ -0,0 +1,107 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {MithrilViewComponent} from "jsx/mithril-component";
+import _ from "lodash";
+import m from "mithril";
+import {Material, Materials} from "models/new_pipeline_configs/materials";
+import {Secondary} from "views/components/buttons";
+import {CollapsiblePanel} from "views/components/collapsible_panel";
+import {IconGroup} from "views/components/icons";
+import {Delete} from "views/components/icons";
+import {MaterialOperations} from "../../create_pipeline_page";
+import styles from "../index.scss";
+import materialStyles from "./index.scss";
+import {AddMaterialModal} from "./modals";
+
+interface Attrs {
+  materials: Materials;
+  materialOperations: MaterialOperations;
+}
+
+export class MaterialsWidget extends MithrilViewComponent<Attrs> {
+  view(vnode: m.Vnode<Attrs>) {
+    let materials = [];
+    let expanded  = true;
+    if (_.isEmpty(vnode.attrs) || _.isEmpty(vnode.attrs.materials) || vnode.attrs.materials.length === 0) {
+      expanded  = true;
+      materials = ["No materials defined"];
+    } else {
+      materials = [<table className={materialStyles.table} data-test-id={"materials-index-table"}>
+        <thead data-test-id="table-header">
+        <tr data-test-id="table-header-row">
+          <th>Material Name</th>
+          <th>Type</th>
+          <th>Url</th>
+          <th></th>
+        </tr>
+        </thead>
+        {this.materialsData(vnode)}
+      </table>];
+    }
+
+    return <CollapsiblePanel header="Materials"
+                             actions={
+                               <Secondary onclick={(e: MouseEvent) => this.addMaterialModal(e, vnode)}
+                                          data-test-id={"add-material-button"}>Add
+                                 Material</Secondary>
+                             }
+                             expanded={expanded}
+                             dataTestId="pipeline-materials-container">
+      {materials}
+    </CollapsiblePanel>;
+  }
+
+  materialsData(vnode: m.Vnode<Attrs>): any {
+    const tbodyContent = Array<any>();
+    vnode.attrs.materials.forEach((material) => {
+      const materialContent = Array<any>();
+      materialContent.push(<td>{material.name()}</td>);
+      materialContent.push(<td>{this.getTypeForDisplay(material.type())}</td>);
+      materialContent.push(<td>{material.materialUrl()}</td>);
+      materialContent.push(<td class={styles.textAlignRight}>
+        <IconGroup>
+          <Delete onclick={() => vnode.attrs.materialOperations.onDelete(material)} title={"delete-material"}
+                  data-test-id={"delete-material-button"}/>
+        </IconGroup>
+      </td>);
+      tbodyContent.push(<tr>{materialContent}</tr>);
```

Comment:
You could avoid all the `array.push` by defining a single element such as:
```
const materialContent =
              <tr>
                <td>{material.name()}</td>
                <td>{this.getTypeForDisplay(material.type())}</td>
                <td>{material.materialUrl()}</td>
                <td class={styles.textAlignRight}>
                  <IconGroup>
                    <Delete onclick={() => vnode.attrs.materialOperations.onDelete(material)} title={"delete-material"}
                            data-test-id={"delete-material-button"}/>
                  </IconGroup>
                </td>
              </tr>;
```
If you used a `map` instead of `forEach`, you could avoid creating the arrays altogether.

Reply:
It's a `set` so used `forEach`.

Reply:
In subsequent PR, we will be getting rid of `NameableSet` for materials. Mainly because name can be blank. `Materials` class will extend an array which will have its customized method to add material. When this change is done, I will change this code to use `map` instead of `forEach`

Reply:
Ok. I'm not a fan of *extending* collection classes, like we do in the Java code. In ES6 and typescript, the collection classes already have some good in-built functionality and there lodash for everything else. 

In other words, a `Collection<T>` is fine, but `T extends Collection` may be overkill in JS.